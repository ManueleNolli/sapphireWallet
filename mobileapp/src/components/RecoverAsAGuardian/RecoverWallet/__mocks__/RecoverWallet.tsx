const RecoverWalletMock = () => <div data-testid="RecoverWallet">RecoverWallet</div>

export default RecoverWalletMock
