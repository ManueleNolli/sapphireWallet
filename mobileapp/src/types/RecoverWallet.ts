export type RecoverWallet = {
  walletAddress: string
  chainID: string
  nonce: string
  wrappedTransaction: string
  signedTransaction: string
  mnemonic: string[]
}

export const RecoverWalletToJSON = (value: RecoverWallet): string => {
  const mnemonic = value.mnemonic.join(' ')
  return JSON.stringify({ ...value, mnemonic })
}

export const JSONToRecoverWallet = (value: string): RecoverWallet => {
  const parsed = JSON.parse(value)
  return { ...parsed, mnemonic: parsed.mnemonic.split(' ') }
}
