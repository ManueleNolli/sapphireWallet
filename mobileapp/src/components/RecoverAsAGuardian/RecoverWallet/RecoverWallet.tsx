import { Button, Layout, Spinner, StyleService, Text } from '@ui-kitten/components'
import { appStyles, vh, vw } from '../../../Styles'
import React from 'react'
import useRecoverWallet from './useRecoverWallet'
import { View } from 'react-native'
import QRCode from 'react-native-qrcode-svg'

type RecoverWalletProps = {
  wallet: string
}

export default function RecoverWallet(props: RecoverWalletProps) {
  const { isLoading, prepareTransactionAndShowQrCode, recoverWalletInfo } = useRecoverWallet(props)

  return (
    <Layout style={styles.modalContainer}>
      <Text category="h6" style={{ marginBottom: 2 * vh }}>
        Recover wallet
      </Text>

      {recoverWalletInfo ? (
        <QRCode value={recoverWalletInfo} size={75 * vw} />
      ) : isLoading ? (
        <View style={appStyles.center}>
          <Spinner
            size="giant"
            status="basic"
            style={{
              marginBottom: 2 * vh,
            }}
          />
        </View>
      ) : (
        <>
          <Text style={{ textAlign: 'center' }}>Are you sure you want to recover wallet</Text>
          <Text
            style={{
              textAlign: 'center',
              marginBottom: 2 * vh,
            }}
            category="label"
          >
            {props.wallet}
          </Text>

          <Button testID="button-yen" style={{ width: '50%' }} onPress={prepareTransactionAndShowQrCode}>
            Yes
          </Button>
        </>
      )}
    </Layout>
  )
}

const styles = StyleService.create({
  modalContainer: {
    maxHeight: 75 * vh,
    minHeight: 25 * vh,
    maxWidth: 90 * vw,
    minWidth: 90 * vw,
    borderRadius: 10,
    paddingVertical: 4 * vh,
    paddingHorizontal: 4 * vw,
    alignItems: 'center',
  },
})
