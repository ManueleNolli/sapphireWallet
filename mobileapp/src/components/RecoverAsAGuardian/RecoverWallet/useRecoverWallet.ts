import { useContext, useState } from 'react'
import useLoading from '../../../hooks/useLoading'
import { RecoverWallet, RecoverWalletToJSON } from '../../../types/RecoverWallet'
import { prepareRecoverWallet } from '../../../services/transactions'
import { createWallet, getSigner } from '../../../services/wallet'
import { WalletContext } from '../../../context/WalletContext'
import { BlockchainContext } from '../../../context/BlockchainContext'

type useRecoverWalletProps = {
  wallet: string
}

export function useRecoverWallet({ wallet }: useRecoverWalletProps) {
  const { getPrivateKey } = useContext(WalletContext)
  const { currentNetwork } = useContext(BlockchainContext)
  const { isLoading, setIsLoading } = useLoading()
  const [recoverWalletInfo, setRecoverWalletInfo] = useState<string | undefined>()

  const prepareTransactionAndShowQrCode = async () => {
    setIsLoading(true)
    const signer = await getSigner(await getPrivateKey('Confirm recovery by signing the transaction'), currentNetwork)
    const newWallet = createWallet()
    const data = await prepareRecoverWallet(signer, currentNetwork, wallet, newWallet)
    const dataString = RecoverWalletToJSON(data)
    setRecoverWalletInfo(dataString)
    setIsLoading(false)
  }

  return {
    isLoading,
    prepareTransactionAndShowQrCode,
    recoverWalletInfo,
  }
}
