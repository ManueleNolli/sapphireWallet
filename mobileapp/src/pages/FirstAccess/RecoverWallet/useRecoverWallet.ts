import { useContext, useState } from 'react'
import { JSONToRecoverWallet, RecoverWallet } from '../../../types/RecoverWallet'
import Toast from 'react-native-toast-message'
import { HDNodeWallet, Mnemonic } from 'ethers'
import useLoading from '../../../hooks/useLoading'
import { WalletContext } from '../../../context/WalletContext'
import { getMnemonic } from '../../../services/wallet'
import { RecoverWalletProps } from '../../../navigation/FirstAccessStack'

export default function useRecoverWallet({ route, navigation }: RecoverWalletProps) {
  const { setPrivateKey, setEOAAddress, setWalletContractAddress } = useContext(WalletContext)
  const [isScannedOpen, setIsScannedOpen] = useState(false)
  const { isLoading, setIsLoading } = useLoading()

  const onQrCodeScanned = async (data: string) => {
    setIsLoading(true)
    closeModal()
    try {
      const dataParsed: RecoverWallet = JSONToRecoverWallet(data)
      const mnemonic = Mnemonic.fromPhrase(dataParsed.mnemonic.join(' '))
      const wallet = HDNodeWallet.fromMnemonic(mnemonic)
      await Promise.all([
        setEOAAddress(wallet.address),
        setPrivateKey(wallet.privateKey),
        setWalletContractAddress(dataParsed.walletAddress),
      ])
      navigation.push('MnemonicViewerRecoverWallet', {
        mnemonic: getMnemonic(wallet),
        data: dataParsed,
      })
    } catch (e) {
      Toast.show({
        type: 'error',
        text1: 'Invalid QR code ðŸš¨',
        text2: 'The QR code you scanned is not valid',
      })
    } finally {
      setIsLoading(false)
    }
  }

  const closeModal = () => setIsScannedOpen(false)
  const openModal = () => setIsScannedOpen(true)

  return {
    isScannedOpen,
    closeModal,
    openModal,
    onQrCodeScanned,
    isLoading,
  }
}
