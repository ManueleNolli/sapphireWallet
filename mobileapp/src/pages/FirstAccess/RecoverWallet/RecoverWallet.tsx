import React from 'react'
import { Button, Layout, Modal, Spinner, Text } from '@ui-kitten/components'
import { RecoverWalletProps } from '../../../navigation/FirstAccessStack'
import SafeAreaView from '../../../utils/SafeAreaView'
import { appStyles, vh } from '../../../Styles'
import QRCodeScanner from '../../../components/QRCodeScanner/QRCodeScanner'
import { useRecoverWallet } from './useRecoverWallet'
import { View } from 'react-native'

export default function RecoverWallet({ route, navigation }: RecoverWalletProps) {
  const { isScannedOpen, closeModal, openModal, onQrCodeScanned, isLoading } = useRecoverWallet({ route, navigation })
  return (
    <Layout style={{ flex: 1 }}>
      <SafeAreaView style={{ flex: 1, justifyContent: 'center' }}>
        <Text category="h4" style={{ marginBottom: 2 * vh, textAlign: 'center' }}>
          Recover wallet
        </Text>
        <Text style={{ textAlign: 'center' }}>Scan the QR code of your Guardian to recover your wallet</Text>

        {isLoading ? (
          <View style={appStyles.center}>
            <Spinner
              size="giant"
              status="basic"
              style={{
                marginBottom: 2 * vh,
              }}
            />
          </View>
        ) : (
          <Button appearance="outline" size="giant" onPress={openModal} style={{ marginTop: 6 * vh }}>
            Scan QR code
          </Button>
        )}

        <Modal
          visible={isScannedOpen}
          backdropStyle={{ backgroundColor: 'color-basic-transparent-600' }}
          onBackdropPress={closeModal}
        >
          <QRCodeScanner onQRCodeScanned={onQrCodeScanned} />
        </Modal>
      </SafeAreaView>
    </Layout>
  )
}
