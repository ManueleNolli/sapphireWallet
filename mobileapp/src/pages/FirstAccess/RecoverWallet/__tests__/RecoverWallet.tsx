import useRecoverWallet from '../useRecoverWallet'
import renderWithTheme from '../../../../TestHelper'
import RecoverWallet from '../RecoverWallet'
import { View } from 'react-native'
import React from 'react'

jest.mock('../useRecoverWallet', () => jest.fn())
jest.mock('../../../../components/QRCodeScanner/QRCodeScanner')
jest.mock('@ui-kitten/components', () => {
  const { View } = require('react-native')
  return {
    ...jest.requireActual('@ui-kitten/components'),
    Spinner: () => <View data-testid="spinner" />,
  }
})
describe('RecoverWallet', () => {
  it('should render correctly', () => {
    ;(useRecoverWallet as jest.Mock).mockReturnValue({
      isScannedOpen: false,
      closeModal: jest.fn(),
      openModal: jest.fn(),
      onQrCodeScanned: jest.fn(),
      isLoading: false,
    })

    const tree = renderWithTheme(<RecoverWallet {...({ route: {}, navigation: {} } as any)} />)
    expect(tree).toMatchSnapshot()
  })

  it('should render correctly when loading', () => {
    ;(useRecoverWallet as jest.Mock).mockReturnValue({
      isScannedOpen: false,
      closeModal: jest.fn(),
      openModal: jest.fn(),
      onQrCodeScanned: jest.fn(),
      isLoading: true,
    })

    const tree = renderWithTheme(<RecoverWallet {...({ route: {}, navigation: {} } as any)} />)
    expect(tree).toMatchSnapshot()
  })

  it('should render correctly when scanning', () => {
    ;(useRecoverWallet as jest.Mock).mockReturnValue({
      isScannedOpen: true,
      closeModal: jest.fn(),
      openModal: jest.fn(),
      onQrCodeScanned: jest.fn(),
      isLoading: false,
    })

    const tree = renderWithTheme(<RecoverWallet {...({ route: {}, navigation: {} } as any)} />)
    expect(tree).toMatchSnapshot()
  })
})
