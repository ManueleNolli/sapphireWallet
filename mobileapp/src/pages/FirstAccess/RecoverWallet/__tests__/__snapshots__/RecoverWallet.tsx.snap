// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecoverWallet should render correctly 1`] = `
<View
  appearance="default"
  level="1"
  style={
    [
      {
        "backgroundColor": "#FFFFFF",
      },
      {
        "flex": 1,
      },
    ]
  }
>
  <View
    appearance="default"
    level="1"
    style={
      [
        {
          "backgroundColor": "#FFFFFF",
        },
        {
          "flex": 1,
        },
      ]
    }
  >
    <RNCSafeAreaView
      edges={
        {
          "bottom": "additive",
          "left": "additive",
          "right": "additive",
          "top": "additive",
        }
      }
      style={
        [
          {
            "flex": 1,
            "marginHorizontal": 37.5,
            "marginVertical": 66.7,
          },
          {
            "flex": 1,
            "justifyContent": "center",
          },
        ]
      }
    >
      <Text
        appearance="default"
        category="h4"
        status="basic"
        style={
          [
            {
              "color": "#222B45",
              "fontFamily": "System",
              "fontSize": 26,
              "fontWeight": "800",
            },
            {
              "marginBottom": 26.68,
              "textAlign": "center",
            },
          ]
        }
      >
        Recover wallet
      </Text>
      <Text
        appearance="default"
        category="p1"
        status="basic"
        style={
          [
            {
              "color": "#222B45",
              "fontFamily": "System",
              "fontSize": 15,
              "fontWeight": "400",
            },
            {
              "textAlign": "center",
            },
          ]
        }
      >
        Scan the QR code of your Guardian to recover your wallet
      </Text>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "rgba(51, 102, 255, 0.08)",
            "borderColor": "#3366FF",
            "borderRadius": 4,
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "center",
            "marginTop": 80.03999999999999,
            "minHeight": 56,
            "minWidth": 56,
            "opacity": 1,
            "paddingHorizontal": 12,
            "paddingVertical": 16,
          }
        }
      >
        <Text
          appearance="default"
          category="p1"
          status="basic"
          style={
            [
              {
                "color": "#222B45",
                "fontFamily": "System",
                "fontSize": 15,
                "fontWeight": "400",
              },
              {
                "color": "#3366FF",
                "fontFamily": "System",
                "fontSize": 18,
                "fontWeight": "bold",
                "marginHorizontal": 12,
              },
            ]
          }
        >
          Scan QR code
        </Text>
      </View>
    </RNCSafeAreaView>
  </View>
</View>
`;

exports[`RecoverWallet should render correctly when loading 1`] = `
<View
  appearance="default"
  level="1"
  style={
    [
      {
        "backgroundColor": "#FFFFFF",
      },
      {
        "flex": 1,
      },
    ]
  }
>
  <View
    appearance="default"
    level="1"
    style={
      [
        {
          "backgroundColor": "#FFFFFF",
        },
        {
          "flex": 1,
        },
      ]
    }
  >
    <RNCSafeAreaView
      edges={
        {
          "bottom": "additive",
          "left": "additive",
          "right": "additive",
          "top": "additive",
        }
      }
      style={
        [
          {
            "flex": 1,
            "marginHorizontal": 37.5,
            "marginVertical": 66.7,
          },
          {
            "flex": 1,
            "justifyContent": "center",
          },
        ]
      }
    >
      <Text
        appearance="default"
        category="h4"
        status="basic"
        style={
          [
            {
              "color": "#222B45",
              "fontFamily": "System",
              "fontSize": 26,
              "fontWeight": "800",
            },
            {
              "marginBottom": 26.68,
              "textAlign": "center",
            },
          ]
        }
      >
        Recover wallet
      </Text>
      <Text
        appearance="default"
        category="p1"
        status="basic"
        style={
          [
            {
              "color": "#222B45",
              "fontFamily": "System",
              "fontSize": 15,
              "fontWeight": "400",
            },
            {
              "textAlign": "center",
            },
          ]
        }
      >
        Scan the QR code of your Guardian to recover your wallet
      </Text>
      <View
        style={
          {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <View
          data-testid="spinner"
        />
      </View>
    </RNCSafeAreaView>
  </View>
</View>
`;

exports[`RecoverWallet should render correctly when scanning 1`] = `
<View
  appearance="default"
  level="1"
  style={
    [
      {
        "backgroundColor": "#FFFFFF",
      },
      {
        "flex": 1,
      },
    ]
  }
>
  <View
    appearance="default"
    level="1"
    style={
      [
        {
          "backgroundColor": "#FFFFFF",
        },
        {
          "flex": 1,
        },
      ]
    }
  >
    <RNCSafeAreaView
      edges={
        {
          "bottom": "additive",
          "left": "additive",
          "right": "additive",
          "top": "additive",
        }
      }
      style={
        [
          {
            "flex": 1,
            "marginHorizontal": 37.5,
            "marginVertical": 66.7,
          },
          {
            "flex": 1,
            "justifyContent": "center",
          },
        ]
      }
    >
      <Text
        appearance="default"
        category="h4"
        status="basic"
        style={
          [
            {
              "color": "#222B45",
              "fontFamily": "System",
              "fontSize": 26,
              "fontWeight": "800",
            },
            {
              "marginBottom": 26.68,
              "textAlign": "center",
            },
          ]
        }
      >
        Recover wallet
      </Text>
      <Text
        appearance="default"
        category="p1"
        status="basic"
        style={
          [
            {
              "color": "#222B45",
              "fontFamily": "System",
              "fontSize": 15,
              "fontWeight": "400",
            },
            {
              "textAlign": "center",
            },
          ]
        }
      >
        Scan the QR code of your Guardian to recover your wallet
      </Text>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "rgba(51, 102, 255, 0.08)",
            "borderColor": "#3366FF",
            "borderRadius": 4,
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "center",
            "marginTop": 80.03999999999999,
            "minHeight": 56,
            "minWidth": 56,
            "opacity": 1,
            "paddingHorizontal": 12,
            "paddingVertical": 16,
          }
        }
      >
        <Text
          appearance="default"
          category="p1"
          status="basic"
          style={
            [
              {
                "color": "#222B45",
                "fontFamily": "System",
                "fontSize": 15,
                "fontWeight": "400",
              },
              {
                "color": "#3366FF",
                "fontFamily": "System",
                "fontSize": 18,
                "fontWeight": "bold",
                "marginHorizontal": 12,
              },
            ]
          }
        >
          Scan QR code
        </Text>
      </View>
      <Modal
        hardwareAccelerated={false}
        onDismiss={[MockFunction]}
        onRequestClose={[MockFunction]}
        statusBarTranslucent={false}
        transparent={true}
        visible={true}
      >
        <View
          style={
            {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "backgroundColor": "color-basic-transparent-600",
                "bottom": 0,
                "left": 0,
                "opacity": 1,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
            testID="@backdrop"
          />
          <View
            backdropStyle={
              {
                "backgroundColor": "color-basic-transparent-600",
              }
            }
            onBackdropPress={[MockFunction]}
            onLayout={[Function]}
            shouldUseContainer={true}
            style={
              [
                undefined,
                {
                  "position": "absolute",
                },
                {
                  "left": -999,
                  "top": -999,
                },
              ]
            }
            visible={true}
          >
            <div
              data-testid="QRCodeScanner"
            >
              QRCodeScanner
            </div>
          </View>
        </View>
      </Modal>
    </RNCSafeAreaView>
  </View>
</View>
`;
